<a href="https://colab.research.google.com/github/salazarcristian1014/restaurant-cancelation-project/blob/main/notebooks/01_data_cleaning_and_optimization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

Autorizamos el acceso de coolab a nuestros archivos de google drive
*We authorize Colab to access our Google Drive files.*

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
# Guardamos la ruta del archivo en una variable
# We authorize Colab to access our Google Drive files.
ruta_archivo = '/content/drive/MyDrive/Reservas_Restaurante/Reservas_Restaurante.csv'

# Leemos el archivo CSV y lo convertimos en nuestro DataFrame llamado 'df'
# We read the CSV file and convert it into our DataFrame named 'df'
df = pd.read_csv(ruta_archivo)

# La función .head() nos muestra las primeras 5 filas para verificar que todo cargó bien
# The .head() function shows us the first 5 rows to verify that everything loaded correctly
df.head()

,codigo,sucursal,fecha,hora,Número de personas,Nuevo usuario,nombre,email,código de teléfono,teléfono,...,FE teléfono,FE billingId,acepta términos y condiciones de Cluvi,fecha de aceptación de términos y condiciones de Cluvi,hora de aceptación de términos y condiciones de Cluvi,términos y condiciones de Cluvi,acepta términos y condiciones del restaurante,fecha de aceptación de términos y condiciones del restaurante,hora de aceptación de términos y condiciones del restaurante,términos y condiciones del restaurante
0,2002222,Restaurante Idilico,2025-01-02,20:30,2,Sí,Francesca Grimaldi,f.schibonigrimaldi@gmail.com,39,3288317828,...,NaN,NaN,Sí,02-12-2024,23:56:58,NaN,Sí,02-12-2024,23:56:58,NaN
1,2025131,Restaurante Idilico,2025-01-04,20:30,2,No,Samantha Vassor,samvassor@gmail.com,1,5616939883,...,NaN,NaN,Sí,16-12-2024,11:26:49,NaN,Sí,16-12-2024,11:26:49,NaN
2,2007791,Restaurante Idilico,2025-01-04,20:30,2,Sí,Melani Kerslake,mkerslake@msn.com,1,4252757101,...,NaN,NaN,Sí,06-12-2024,06:06:04,NaN,Sí,06-12-2024,06:06:04,NaN
3,2008810,Restaurante Idilico,2025-01-06,18:30,1,Sí,Michael Lowe,michael.lowe@tefball.com,1,7783503113,...,NaN,NaN,Sí,06-12-2024,12:56:52,NaN,Sí,06-12-2024,12:56:52,NaN
4,2009405,Restaurante Idilico,2025-01-06,18:30,2,Sí,Michelle yu,michelleyu94@gmail.com,1,4168785661,...,NaN,NaN,Sí,06-12-2024,19:46:18,NaN,Sí,06-12-2024,19:46:18,NaN


Revisamos que tipo de columnas tenemos.
*We check what type of columns we have.*




In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3219 entries, 0 to 3218
Data columns (total 52 columns):
 #   Column                                                         Non-Null Count  Dtype  
---  ------                                                         --------------  -----  
 0   codigo                                                         3219 non-null   int64  
 1   sucursal                                                       3219 non-null   object 
 2   fecha                                                          3219 non-null   object 
 3   hora                                                           3219 non-null   object 
 4   Número de personas                                             3219 non-null   int64  
 5   Nuevo usuario                                                  3219 non-null   object 
 6   nombre                                                         3219 non-null   object 
 7   email                                                       

Procedemos a eliminar las columnas que no tienen valores en ninguna de sus filas y luego revisamos que si se hayan eliminado.

*We proceed to drop the columns that have no values in any of their rows, and then we verify that they have been removed.*

In [5]:
df=df.dropna(axis = 1, how='all')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3219 entries, 0 to 3218
Data columns (total 39 columns):
 #   Column                                                         Non-Null Count  Dtype 
---  ------                                                         --------------  ----- 
 0   codigo                                                         3219 non-null   int64 
 1   sucursal                                                       3219 non-null   object
 2   fecha                                                          3219 non-null   object
 3   hora                                                           3219 non-null   object
 4   Número de personas                                             3219 non-null   int64 
 5   Nuevo usuario                                                  3219 non-null   object
 6   nombre                                                         3219 non-null   object
 7   email                                                          3105 n

Eliminamos desde la columna 26 hasta la columna 38 ya que estas columnas nos aportan datos de facturacion electronica y permisos legales que solicita el sistema de reservas y esta informacion no es util para el problema que queremos resolver

*We drop columns 26 through 38, as these columns provide electronic billing data and legal permissions requested by the reservation system, and this information is not useful for the problem we want to solve.*

In [6]:
#Revisamos los nombres de las columnas para copiarlos y pegarlos de una manera mas practica en la proxima linea de codigo
# We check the column names so we can copy and paste them more practically in the next line of code
df.columns[26:39]

Index(['FE documento', 'FE Nombre', 'FE email', 'FE teléfono', 'FE billingId',
       'acepta términos y condiciones de Cluvi',
       'fecha de aceptación de términos y condiciones de Cluvi',
       'hora de aceptación de términos y condiciones de Cluvi',
       'términos y condiciones de Cluvi',
       'acepta términos y condiciones del restaurante',
       'fecha de aceptación de términos y condiciones del restaurante',
       'hora de aceptación de términos y condiciones del restaurante',
       'términos y condiciones del restaurante'],
      dtype='object')

In [7]:
#Eliminamos las columnas con informacion legal y datos de facturacion electronica
# We drop the columns containing legal information and electronic billing data
df = df.drop(columns =['FE documento', 'FE Nombre', 'FE email', 'FE teléfono', 'FE billingId',
       'acepta términos y condiciones de Cluvi',
       'fecha de aceptación de términos y condiciones de Cluvi',
       'hora de aceptación de términos y condiciones de Cluvi',
       'términos y condiciones de Cluvi',
       'acepta términos y condiciones del restaurante',
       'fecha de aceptación de términos y condiciones del restaurante',
       'hora de aceptación de términos y condiciones del restaurante',
       'términos y condiciones del restaurante'] )

In [8]:
# Revisamos nuevamente como va la limpieza de nuestro archivo
# We check again how our file cleaning is progressing
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3219 entries, 0 to 3218
Data columns (total 26 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   codigo                   3219 non-null   int64 
 1   sucursal                 3219 non-null   object
 2   fecha                    3219 non-null   object
 3   hora                     3219 non-null   object
 4   Número de personas       3219 non-null   int64 
 5   Nuevo usuario            3219 non-null   object
 6   nombre                   3219 non-null   object
 7   email                    3105 non-null   object
 8   código de teléfono       3218 non-null   object
 9   teléfono                 3219 non-null   int64 
 10  estado                   3219 non-null   object
 11  observación              146 non-null    object
 12  Comentario Reserva       427 non-null    object
 13  Observación al cancelar  28 non-null     object
 14  Tipo de reserva          3219 non-null  

Eliminamos las columnas de Alergia, responsable, llego y observacion ya que tienen muy pocos datos y ademas no estan relacionados con temas de cancelaciones y revisamos nuevamente como va nuestro archivo.
*We drop the 'Alergia' (Allergy), 'responsable' (Responsible), 'llego' (Arrived), and 'observacion' (Observation) columns since they have very little data and are not related to cancellation issues, and we check our file again.*

In [9]:
df = df.drop(columns = ['Alergias', 'Responsable', 'llego', 'observación'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3219 entries, 0 to 3218
Data columns (total 22 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   codigo                   3219 non-null   int64 
 1   sucursal                 3219 non-null   object
 2   fecha                    3219 non-null   object
 3   hora                     3219 non-null   object
 4   Número de personas       3219 non-null   int64 
 5   Nuevo usuario            3219 non-null   object
 6   nombre                   3219 non-null   object
 7   email                    3105 non-null   object
 8   código de teléfono       3218 non-null   object
 9   teléfono                 3219 non-null   int64 
 10  estado                   3219 non-null   object
 11  Comentario Reserva       427 non-null    object
 12  Observación al cancelar  28 non-null     object
 13  Tipo de reserva          3219 non-null   object
 14  origen                   3219 non-null  

Aqui teniendo en cuenta que ya sacamos al informacion basura que no tenia datos, columnas que no nos servian para el analisis y columnas con muy pocos datos... Implementamos algo de logica para el logro principal de este proyecto que es conocer cual es la tasa de cancelacion en base al total de reservas por esta razon vamos a revisar los datos de cada una de las columnas y miramos si la dejamos o la descartamos.

*Here, considering that we have already removed the junk information that had no data, columns that were useless for the analysis, and columns with very little data... We implement some logic for the main goal of this project, which is to determine the cancellation rate based on total reservations. For this reason, we will review the data in each of the columns and see if we keep it or discard it.*

In [10]:
#El codigo lo dejamos ya que nos sirve como contador de numero total de reservas
# We keep the 'codigo' (code) column as it serves as a counter for the total number of reservations
df['codigo'].value_counts()

,count
codigo,
2770837,1
2002222,1
2025131,1
2774424,1
2774473,1
...,...
2037213,1
2057467,1
2048525,1


In [11]:
#La sucursal la dejamos ya que nos puede ayudar a segmentar los datos y conocer en que sucursal se presentan mayor numero de cancelaciones
# We keep the 'sucursal' (branch) column as it can help us segment the data and find out which branch has the highest number of cancellations
df['sucursal'].value_counts()


,count
sucursal,
Restaurante Idilico,2773
Mesas Del Chef,446


In [12]:
#Con este codigo podemos ver desde y hasta que fecha tenemos los datos, por esa razon vamos a organizar
# el formato para que no se lea como object sino como datetime
# With this code, we can see the date range of our data; for that reason, we are going to format
# it so it is not read as an 'object' but as 'datetime'
df['fecha'].value_counts()

,count
fecha,
2026-02-14,38
2026-01-24,34
2025-02-21,32
2026-02-13,30
2026-02-21,25
...,...
2025-12-25,1
2025-04-24,1
2025-04-30,1


In [13]:
# Convertimos la columna fecha a tipo datetime
# We convert the 'fecha' (date) column to datetime type
df['fecha'] = pd.to_datetime(df['fecha'])

In [14]:
# La columna hora la dejaremos ya que nos puede dar datos de cuales son los horarios mas congestionados y al igual que con la fecha
# la convertimos de dato object a datetime
# We will keep the 'hora' (time) column as it can provide data on the busiest hours, and just like with the date,
# we convert it from 'object' data type to 'datetime'
df['hora'] = pd.to_datetime(df['hora'])
df['hora'].value_counts()

/tmp/ipykernel_2171/898922266.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['hora'] = pd.to_datetime(df['hora'])


,count
hora,
2026-03-21 20:30:00,1050
2026-03-21 18:30:00,811
2026-03-21 19:30:00,389
2026-03-21 19:00:00,287
2026-03-21 13:00:00,263
2026-03-21 12:00:00,139
2026-03-21 21:00:00,130
2026-03-21 20:00:00,107
2026-03-21 19:15:00,12


In [15]:
# Esta columna es vital para calcular el numero de personas que Reservaron ya que pueden haber muchas reservas registradas pero
# solo conociendo el numero de personas por reserva podremos saber en total cuantas personas reservaron.
# This column is vital for calculating the number of people who booked, since there might be many registered reservations, but
# only by knowing the number of people per reservation will we be able to know the total number of people who booked.
df['Número de personas'].value_counts()

,count
Número de personas,
2,1875
4,413
3,351
1,236
6,180
5,111
8,14
7,13
10,8


In [16]:
#Esta variable nos puede ayudar a entender si el porcentaje de cancelaciones es mayor en nuevas personas o en personas que ya nos han
# This variable can help us understand if the cancellation percentage is higher among new guests or returning customers
df['Nuevo usuario'].value_counts()

,count
Nuevo usuario,
Sí,2755
No,464


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3219 entries, 0 to 3218
Data columns (total 22 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   codigo                   3219 non-null   int64         
 1   sucursal                 3219 non-null   object        
 2   fecha                    3219 non-null   datetime64[ns]
 3   hora                     3219 non-null   datetime64[ns]
 4   Número de personas       3219 non-null   int64         
 5   Nuevo usuario            3219 non-null   object        
 6   nombre                   3219 non-null   object        
 7   email                    3105 non-null   object        
 8   código de teléfono       3218 non-null   object        
 9   teléfono                 3219 non-null   int64         
 10  estado                   3219 non-null   object        
 11  Comentario Reserva       427 non-null    object        
 12  Observación al cancelar  28 non-nu

Las columnas con nombre, email, codigo de telefono y telefono a pesar de que no las necesitamos para el analisis inicial sobre la tasa de cancelacion son de vital importancia para la segunda parte del proyecto donde buscaremos al forma de automatizar esas reservas canceladas e invitarlas a que nos digan el motivo de la cancelacion y ademas poder reinvitarlos al restaurante.

*The columns containing name, email, country code, and phone number, although not needed for the initial cancellation rate analysis, are of vital importance for the second part of the project. There, we will look for a way to automate outreach to those canceled reservations, inviting them to tell us the reason for the cancellation, and also being able to invite them back to the restaurant.*

In [18]:
# Procedemos a modificar el tipo de dato int de la columna 'telefono' para luego poder concatenarlo con 'codigo' y al momento
# de la automatizacion se puedan enviar mensajes por whatsapp
# We proceed to modify the 'int' data type of the 'telefono' (phone) column so we can later concatenate it with 'codigo' (country code), so that at the time
# of the automation, we can send messages via WhatsApp
df['teléfono'] = df['teléfono'].astype(str)

In [19]:
# Creamos la columna 'telefono_completo' esto con el fin de concatenar el codigo y el telefono para poder enviar mensajes via whatsapp
# We create the 'telefono_completo' (full phone) column in order to concatenate the country code and the phone number to be able to send messages via WhatsApp
df['telefono_completo'] = '+' + df['código de teléfono'] + df['teléfono']

In [20]:
df['telefono_completo'].value_counts()

,count
telefono_completo,
+573014954440,17
+19543059444,13
+573146484974,11
+4540330364,10
+491621768928,9
...,...
+573003753957,1
+573164782185,1
+573187825762,1


In [21]:
# Procedemos a eliminar la columna 'telefono' ya que al tener el 'telefono_completo' no es necesaria tenerla en nuestro df, dejamos la
# columna de codigo ya que nos puede servir para identificar de que paises nos visitan en su mayoria.
# We proceed to drop the 'telefono' column since having 'telefono_completo' makes it unnecessary in our df. We keep the
# 'codigo' column as it can help us identify which countries the majority of our visitors come from.
df = df.drop(['teléfono'], axis = 1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3219 entries, 0 to 3218
Data columns (total 22 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   codigo                   3219 non-null   int64         
 1   sucursal                 3219 non-null   object        
 2   fecha                    3219 non-null   datetime64[ns]
 3   hora                     3219 non-null   datetime64[ns]
 4   Número de personas       3219 non-null   int64         
 5   Nuevo usuario            3219 non-null   object        
 6   nombre                   3219 non-null   object        
 7   email                    3105 non-null   object        
 8   código de teléfono       3218 non-null   object        
 9   estado                   3219 non-null   object        
 10  Comentario Reserva       427 non-null    object        
 11  Observación al cancelar  28 non-null     object        
 12  Tipo de reserva          3219 non-

Teniendo en cuenta que la columna que creamos telefono_completo quedo al final y lo ideal es que quede junto con la informacion personal vamos a organizar de tal forma que se pueda mover y quede antes de 'estado'

*Considering that the 'telefono_completo' column we created ended up at the end, and ideally it should be alongside the personal information, we are going to reorganize it so it can be moved right before 'estado' (status).*

In [22]:
#revisamos las columnas que tenemos actualmente
# We check the columns we currently have
df.columns


Index(['codigo', 'sucursal', 'fecha', 'hora', 'Número de personas',
       'Nuevo usuario', 'nombre', 'email', 'código de teléfono', 'estado',
       'Comentario Reserva', 'Observación al cancelar', 'Tipo de reserva',
       'origen', 'salió', 'Fecha de creación', 'Hora de creación',
       'Fecha de cancelación', 'Hora de cancelación', 'Cobros', 'Pagado',
       'telefono_completo'],
      dtype='object')

In [23]:
# teniendo en cuenta que la columna que creamos con 'telefono_completo' quedo al final y con el fin de hacerlo de una forma mas organizada
# procedemos a reestructurar el orden de las columnas de tal forma que la informacion personal quede lo mas junta posible.
# Considering that the column we created with 'telefono_completo' ended up at the end, and in order to make it more organized,
# we proceed to restructure the column order so that the personal information is grouped as closely together as possible.
nuevo_orden = [
    'codigo', 'sucursal', 'fecha', 'hora', 'Número de personas', 'Nuevo usuario',
    'nombre', 'email', 'código de teléfono', 'telefono_completo', 'estado',
    'Comentario Reserva', 'Observación al cancelar', 'Tipo de reserva', 'origen', 'salió',
    'Fecha de creación', 'Hora de creación', 'Fecha de cancelación', 'Hora de cancelación',
    'Cobros', 'Pagado'
]

# Aplicamos el nuevo orden al DataFrame
# We apply the new order to the DataFrame
df = df[nuevo_orden]

# Verificamos como va nuestro archivo
# We verify how is our file dateframe going
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3219 entries, 0 to 3218
Data columns (total 22 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   codigo                   3219 non-null   int64         
 1   sucursal                 3219 non-null   object        
 2   fecha                    3219 non-null   datetime64[ns]
 3   hora                     3219 non-null   datetime64[ns]
 4   Número de personas       3219 non-null   int64         
 5   Nuevo usuario            3219 non-null   object        
 6   nombre                   3219 non-null   object        
 7   email                    3105 non-null   object        
 8   código de teléfono       3218 non-null   object        
 9   telefono_completo        3218 non-null   object        
 10  estado                   3219 non-null   object        
 11  Comentario Reserva       427 non-null    object        
 12  Observación al cancelar  28 non-nu

Hasta este punto ya hemos limpiado desde la columna codigo hasta la columna 'telefono_completo', Lo que sigue la proxima vez que inicie sesion es continuar con la limpieza desde la columna estado hasta la columna pagados.

*Up to this point, we have cleaned everything from the 'codigo' column to the 'telefono_completo' column. The next step the next time I log in is to continue the cleaning process from the 'estado' column to the 'pagados' (paid) column.*

In [24]:
# utilizar este codigo para ir revisando los valores de cada columna e ir viendo que se debe limpiar
# Use this code to keep checking the values of each column and seeing what needs to be cleaned
df['estado'].value_counts()

,count
estado,
Re Confirmada,1367
"{re_confirmed: ""Re Confirmada"", confirmed: ""Confirmada"", wait_list: ""Lista de espera"", pending: ""Pendiente"", sit: ""Sentado"", no_show: ""El cliente no llegó a la reserva"", cancelled: ""Cancelada"", finished: ""Finalizada"", annulled: ""Anulada""}",1060
Confirmada,388
Cancelada,362
Pendiente,20
Translation missing: es.reservation.status_code.unknown,10
Anulada,9
El cliente no llegó a la reserva,2
Sentado,1


Vaya, al parecer aqui tenemos mucho por limpiar ya que tenemos tenemos mas de 1060 datos que no tienen un estado definido, lo que haremos sera lo siguiente:

1. Vamos a manejar solo 2 estados con el fin de hacer la informacion mas facil de entender para el dueño del restaurante. (confirmada y cancelada)
2. las reservas en estado "reconfirmada" , "confirmada" y "sentado" pasaran a ser solo "confirmada"
3. las reservas en estado "el cliente no llego a la la reserva" , "anulada", "pendiente" y "cancelada" pasaran a estas en estado cancelada.
4. Aqui es donde esta el gran problema y es que casi la mitad de nuestra base de datos tiene el siguiente estado el cual parece una falla del sistema "{re_confirmed: "Re Confirmada", confirmed: "Confirmada", wait_list: "Lista de espera", pending: "Pendiente", sit: "Sentado", no_show: "El cliente no llegó a la reserva", cancelled: "Cancelada", finished: "Finalizada", annulled: "Anulada"}" y otro mensaje del sistema que es "{Translation missing: es.reservation.status_code.unknown}"
Lo que haremos sera revisar las otras columnas ( "observacion al cancelar" , "fecha de cancelacion" y "hora de cancelacion") para tratar de intuir cuales de esas reservas cancelaron y cambiarles el estado a "cancelado" y con el resto de reservas pasarlas a estado "confirmada".

*Wow, it looks like we have a lot of cleaning up to do here, as we have over 1,060 records without a defined status. Here is what we are going to do:*

*We will handle only 2 statuses in order to make the information easier to understand for the restaurant owner (confirmed and cancelled).*

*Reservations with the status "reconfirmed", "confirmed", and "seated" will become just "confirmed".*

*Reservations with the status "no-show", "annulled", "pending", and "cancelled" will be changed to "cancelled".*

*Here is where the big problem lies: almost half of our database has the following status, which looks like a system glitch: "{re_confirmed: "Re Confirmada", confirmed: "Confirmada", wait_list: "Lista de espera", pending: "Pendiente", sit: "Sentado", no_show: "El cliente no llegó a la reserva", cancelled: "Cancelada", finished: "Finalizada", annulled: "Anulada"}", alongside another system message that reads "{Translation missing: es.reservation.status_code.unknown}".*

What we will do is review the other columns ("cancellation remarks", "cancellation date", and "cancellation time") to try to infer which of those reservations were cancelled, change their status to "cancelled", and move the rest of the reservations to the "confirmed" status. *texto en cursiva* *texto en cursiva*

In [25]:
#Antes de iniciar con la limpieza que comentamos en el paso anterior vamos a cambiar los datos que ya tenemos a solo 2 opciones ("Confirmada y Cancelada")
#Before we begin the cleanup discussed in the previous step, we will change the existing data to just 2 options ("Confirmed" and "Cancelled").
df['estado']=df['estado'].replace({'Re Confirmada': 'Confirmada' , 'Sentado' : 'Confirmada' ,'El cliente no llegó a la reserva' : 'Cancelada' , 'Anulada' : 'Cancelada' , 'Pendiente' : 'Cancelada'})
df['estado'].value_counts()

,count
estado,
Confirmada,1756
"{re_confirmed: ""Re Confirmada"", confirmed: ""Confirmada"", wait_list: ""Lista de espera"", pending: ""Pendiente"", sit: ""Sentado"", no_show: ""El cliente no llegó a la reserva"", cancelled: ""Cancelada"", finished: ""Finalizada"", annulled: ""Anulada""}",1060
Cancelada,393
Translation missing: es.reservation.status_code.unknown,10


In [26]:
#Vamos a tomar la columna de fecha de cancelacion y donde hayan valores osea "true" procederemos a cambiar el estado de ese dato por "Cancelada"
# We will take the cancellation date column, and where there are values (meaning "true"), we will proceed to change the status of that data to "Cancelled".
df['Fecha de cancelación'].notna()
df.loc[df['Fecha de cancelación'].notna() , 'estado' ] = 'Cancelada'

In [27]:
df['estado'].value_counts()

,count
estado,
Confirmada,1754
"{re_confirmed: ""Re Confirmada"", confirmed: ""Confirmada"", wait_list: ""Lista de espera"", pending: ""Pendiente"", sit: ""Sentado"", no_show: ""El cliente no llegó a la reserva"", cancelled: ""Cancelada"", finished: ""Finalizada"", annulled: ""Anulada""}",867
Cancelada,598


In [28]:
#Antes de repetir el paso anterior con la columna observacion al cancelar revisamos que tipo de datos tenemos en la columna
# Before repeating the previous step with the cancellation remarks column, we will review the type of data we have in the column.
df['Observación al cancelar'].value_counts()

,count
Observación al cancelar,
cliente,17
no abriremos ese dia,2
cancelo via whatsapp,2
CANCELADA por cliente,1
DUPLICADO,1
reserva doble,1
menu degustacion,1
Cambio de horario,1
cancelo,1


In [29]:
# Al parecer aqui tendremos que hacer una limpieza dentro de una limpieza ya que tenemos algunas reservas dobles. Vamos a proceder a eliminar
# las cancelaciones por reservas dobles, lo primero que haremos sera crear una lista con las reservas que queremos borrar
# y que nuestra columna "observaciones al cancelar" nos muestra que son reservas dobles.
# It seems we will have to do a cleanup within a cleanup here, as we have some double reservations. We will proceed to remove
# the cancellations due to double reservations; the first thing we will do is create a list of the reservations we want to delete,
# which our "cancellation remarks" column shows as double reservations.
textos_a_borrar = ['no abriremos ese dia', 'DUPLICADO', 'reserva doble', 'doble']

In [30]:
# Guardamos la condición en una variable
# Save the condition in a variable
condicion_errores = df['Observación al cancelar'].isin(textos_a_borrar)

# Sobrescribimos el DataFrame quedándonos con lo que NO (~) cumple la condición
# Overwrite the DataFrame, keeping only what DOES NOT (~) meet the condition.
df = df[~condicion_errores]

In [31]:
df['Observación al cancelar'].value_counts()

,count
Observación al cancelar,
cliente,17
cancelo via whatsapp,2
CANCELADA por cliente,1
menu degustacion,1
Cambio de horario,1
cancelo,1


In [32]:
#procedemos a cambiar a estado "cancelada" todas las reservas que tienen datos en la columna de 'Observación al cancelar'
# We proceed to change the status to "cancelled" for all reservations that have data in the 'Cancellation note' column

df.loc[df['Observación al cancelar'].notna() , 'estado' ] = 'Cancelada'
df['Observación al cancelar'].value_counts()

,count
Observación al cancelar,
cliente,17
cancelo via whatsapp,2
CANCELADA por cliente,1
menu degustacion,1
Cambio de horario,1
cancelo,1


In [33]:
#Luego de hacer la correcion en base a l fecha de cancelacion y obervacion al cancelar procedemos a cambiar el resto de reservas que
# muestran mensaje de error a "confirmada"
# After making the correction based on the cancellation date and cancellation note
# we proceed to change the remaining reservations showing an error message to "confirmed"

# Creamos la máscara de los que ya están bien
# We create the mask for the ones that are already correct
estados_raros = df['estado'].isin(['Confirmada', 'Cancelada'])

# Reemplazamos los que NO están bien (los errores) por 'Confirmada'
# We replace the ones that are NOT correct (the errors) with 'Confirmed'
df.loc[~estados_raros, 'estado'] = 'Confirmada'
df['estado'].value_counts()

,count
estado,
Confirmada,2616
Cancelada,598


In [34]:
# Teniendo en cuenta que la columna 'Observación al cancelar' ya la usamos para cruzar los datos con 'estado' y que solo tiene 23 datos
# de mas de 3.000 reservas en total vamos a proceder a eliminarla

df = df.drop(columns=['Observación al cancelar'])

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3214 entries, 0 to 3218
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   codigo                3214 non-null   int64         
 1   sucursal              3214 non-null   object        
 2   fecha                 3214 non-null   datetime64[ns]
 3   hora                  3214 non-null   datetime64[ns]
 4   Número de personas    3214 non-null   int64         
 5   Nuevo usuario         3214 non-null   object        
 6   nombre                3214 non-null   object        
 7   email                 3100 non-null   object        
 8   código de teléfono    3213 non-null   object        
 9   telefono_completo     3213 non-null   object        
 10  estado                3214 non-null   object        
 11  Comentario Reserva    427 non-null    object        
 12  Tipo de reserva       3214 non-null   object        
 13  origen                3

In [36]:
#Continuarmos revisando el resto de columnas. Aqui enocntramos algo que tambien se debe limpiar ya que nos puede servir para una
#estrategia de mercadeo donde podamos invitar a las personas a celebrar su cumpleaños en el restaurante, pero hay demasiados datos
# asi que vamos a cambiar la informacion a 2 opciones que son (cumpleaños y aniversario).
df['Comentario Reserva'].value_counts()

,count
Comentario Reserva,
Cumpleaños,19
Cumpleaños,10
Celebración de cumpleaños,6
We are on our honeymoon! :),5
No,5
...,...
luna de miel,1
Celebrando un cumpleaños,1
"En realidad somos 8, es posible?",1


In [37]:
# Aqui convertimos las posibles opciones de personas que han ido a celebrar un cumpleaños al restaurante  en una sola que es 'Cumpleaños'

condicion_cumple = df['Comentario Reserva'].str.contains('cumpleaños|cumple|birthday', case=False, na=False)
df.loc[condicion_cumple, 'Comentario Reserva'] = 'Cumpleaños'
df['Comentario Reserva'].value_counts()

,count
Comentario Reserva,
Cumpleaños,118
We are on our honeymoon! :),5
We’re on our honeymoon! :),5
No,5
"Alergia grave al cacahuete, la avellana y las nueces.",4
...,...
"Hola \nNo sé si sirva de algo escribirlo acá, pero 2 de los comensales no quieren comer pescado y/o mariscos en exceso; primero porque lo disfrutan pero en pequeñas cantidades y segundo porque vendrán de un viaje por la costa caribe y probablemente ya habrán comido buenas cantidades. \nNo sé si es muy atrevido escribir esto, pero lo hago en todo caso. \nMuchas gracias",1
"Hola\nNo sé si sirva de algo escribirlo acá, pero 2 de los comensales no quieren comer pescado y/o mariscos en exceso; primero porque lo disfrutan pero en pequeñas cantidades y segundo porque vendrán de un viaje por la costa caribe y probablemente ya habrán comido buenas cantidades. \nNo sé si es muy atrevido escribir esto, pero lo hago en todo caso.\nMuchas gracias",1
Celebración grados,1


In [38]:
#Vamos a repetir el paso anterior pero con reservas que van a celebrar "Aniversario" o "Anniversary"

condicion_aniversario = df['Comentario Reserva'].str.contains('aniversario|anniversary|aniversary|aniversa', case=False, na=False)
df.loc[condicion_aniversario, 'Comentario Reserva'] = 'Aniversario'
df['Comentario Reserva'].value_counts()

,count
Comentario Reserva,
Cumpleaños,118
Aniversario,31
We are on our honeymoon! :),5
We’re on our honeymoon! :),5
No,5
...,...
Necesitamos opciones vegetarianas para una persona.\n,1
Tienen un libro de vinos que nos puedan enviar por anticipado?,1
Reserva para almuerzo de Mastercard. Esperamos los invitados a las 12:30.,1


In [39]:
# Luego de tener los datos de "Cumpleaños" y "Aniversario" que nos puede servir para estrategias de mercadeo vamos a eliminar
# el resto de comentarios de la columna 'Comentario Reserva'

eventos_clave = df['Comentario Reserva'].isin(['Cumpleaños', 'Aniversario'])
df.loc[~eventos_clave, 'Comentario Reserva'] = None
df['Comentario Reserva'].value_counts()

,count
Comentario Reserva,
Cumpleaños,118
Aniversario,31


In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3214 entries, 0 to 3218
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   codigo                3214 non-null   int64         
 1   sucursal              3214 non-null   object        
 2   fecha                 3214 non-null   datetime64[ns]
 3   hora                  3214 non-null   datetime64[ns]
 4   Número de personas    3214 non-null   int64         
 5   Nuevo usuario         3214 non-null   object        
 6   nombre                3214 non-null   object        
 7   email                 3100 non-null   object        
 8   código de teléfono    3213 non-null   object        
 9   telefono_completo     3213 non-null   object        
 10  estado                3214 non-null   object        
 11  Comentario Reserva    149 non-null    object        
 12  Tipo de reserva       3214 non-null   object        
 13  origen                3

In [42]:
df['Tipo de reserva'].value_counts()

,count
Tipo de reserva,
Reserva,3214


In [43]:
# Esta columna no nos da mayor informacion ya que toda la columna tiene el mismo tipo de dato asi que procedemos a eliminarla

df = df.drop(columns=['Tipo de reserva'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3214 entries, 0 to 3218
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   codigo                3214 non-null   int64         
 1   sucursal              3214 non-null   object        
 2   fecha                 3214 non-null   datetime64[ns]
 3   hora                  3214 non-null   datetime64[ns]
 4   Número de personas    3214 non-null   int64         
 5   Nuevo usuario         3214 non-null   object        
 6   nombre                3214 non-null   object        
 7   email                 3100 non-null   object        
 8   código de teléfono    3213 non-null   object        
 9   telefono_completo     3213 non-null   object        
 10  estado                3214 non-null   object        
 11  Comentario Reserva    149 non-null    object        
 12  origen                3214 non-null   object        
 13  salió                 3

In [47]:
# Aqui vemos el origen por donde reservan las personas, por lo que el google-error que nos muestra lo vamos a corregir y simplemente dejaremos google
df['origen'].value_counts()

,count
origen,
Cluvi,1958
google,1090
Admin,164
google-error,2


In [48]:
# Procedemos a cambiar las reservas que nos muestran error en google por simplemente google

df.loc[df['origen'] == 'google-error', 'origen'] = 'google'
df['origen'].value_counts()

,count
origen,
Cluvi,1958
google,1092
Admin,164


In [50]:
# Esta columna no nos brinda mayor informacion ya que no marcan el momento en que sale la mesa por lo cual procedemos a eliminarla
df['salió'].value_counts()

,count
salió,
No,3214


In [51]:
df=df.drop(columns=['salió'])

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3214 entries, 0 to 3218
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   codigo                3214 non-null   int64         
 1   sucursal              3214 non-null   object        
 2   fecha                 3214 non-null   datetime64[ns]
 3   hora                  3214 non-null   datetime64[ns]
 4   Número de personas    3214 non-null   int64         
 5   Nuevo usuario         3214 non-null   object        
 6   nombre                3214 non-null   object        
 7   email                 3100 non-null   object        
 8   código de teléfono    3213 non-null   object        
 9   telefono_completo     3213 non-null   object        
 10  estado                3214 non-null   object        
 11  Comentario Reserva    149 non-null    object        
 12  origen                3214 non-null   object        
 13  Fecha de creación     3

In [53]:
df['Fecha de creación'].value_counts()

,count
Fecha de creación,
09-02-2026,43
08-03-2025,32
23-01-2026,31
21-02-2025,28
24-01-2026,26
...,...
06-04-2025,1
08-12-2025,1
02-03-2025,1
